src/virtual_machine.rs

```diff
@@ -176,29 +186,165 @@ impl VM {
-fn check_memory_len() {
-    let vm = VM::new();
+    fn update_flags(&mut self, value: u16) {
+        self.rcond = match value.cmp(&0) {
+            Ordering::Less => FL::NEG,
+            Ordering::Equal => FL::ZRO,
+            Ordering::Greater => FL::POS,
+        };
+    }
+
+    fn fetch_value_from_register(&self, reg_num: u16) -> u16 {
+        match reg_num {
+            0 => self.r0,
+            1 => self.r1,
+            2 => self.r2,
+            3 => self.r3,
+            4 => self.r4,
+            5 => self.r5,
+            6 => self.r6,
+            7 => self.r7,
+            8 => self.rpc,
+            10 => self.rcount,
+            _ => panic!("Invalid register"),
+        }
+    }
+
+    fn fetch_register(&mut self, reg_num: u16) -> &mut u16 {
+        match reg_num {
+            0 => &mut self.r0,
+            1 => &mut self.r1,
+            2 => &mut self.r2,
+            3 => &mut self.r3,
+            4 => &mut self.r4,
+            5 => &mut self.r5,
+            6 => &mut self.r6,
+            7 => &mut self.r7,
+            8 => &mut self.rpc,
+            10 => &mut self.rcount,
+            _ => panic!("Invalid register"),
+        }
+    }
 
-    assert_eq!(vm.memory.len(), 65536);
+    fn wrapping_add(a: u16, b: u16) -> u16 {
+        let result = a + b;
+        result
+    }
+
+    fn execute(&mut self, operation: Opcode) {
+        match operation {
+            Opcode::OP_ADD {
+                dr,
+                sr1,
+                second_arg,
+            } => {
+                let sr1_val = self.fetch_value_from_register(sr1);
+                let second_value = match second_arg {
+                    AddMode::IMMEDIATE { imm5 } => {
+                        let value = sign_extend(imm5, 5);
+                        value
+                    }
+                    AddMode::REGISTER { sr2 } => {
+                        let sr2_val = self.fetch_value_from_register(sr2);
+                        sr2_val
+                    }
+                };
+                let result = second_value.wrapping_add(sr1_val);
+
+                let destination = self.fetch_register(dr);
+                *destination = result;
+
+                self.update_flags(result);
+            }
+            _ => todo!(),
+        }
+    }
 }
 
-#[test]
-fn check_memory_add_operation_reg_mode() {
-    let vm = VM::new();
+mod test {
+    use super::*;
 
-    //         ADD R2, R3, R1
-    let op = 0b0001010011000001;
-    let result = VM::decode_instruction(op);
+    #[test]
+    fn check_memory_len() {
+        let vm = VM::new();
 
-    assert_eq!(
-        Opcode::OP_ADD {
-            dr: 2,
-            sr1: 3,
-            second_arg: AddMode::IMMEDIATE { sr2: 1 },
-        },
-        result
-    );
+        assert_eq!(vm.memory.len(), 65536);
+    }
+
+    #[test]
+    fn check_memory_add_operation_reg_mode() {
+        let vm = VM::new();
+
+        //         ADD R2, R3, R1
+        let op = 0b0001010011000001;
```

Comment:
I would separate with `_` so it's clearer to see and less error-prone, same in the remaining ones.